pymeasure/instruments/hp/hplegacyinstrument.py

```diff
@@ -0,0 +1,215 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2021 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+import ctypes
+import logging
+import struct
+from pymeasure.instruments import Instrument
+
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+c_uint8 = ctypes.c_uint8
+c_uint16 = ctypes.c_uint16
+c_uint32 = ctypes.c_uint32
+
+
+class HPLegacyInstrument(Instrument):
```

Comment:
@LongnoseRob, I think your idea of putting the common part in this file is great. If you don't mind, let me share few idea to increase code reuse. I think that status register is in some way pivotal in this instruments, as such I would like to propose the following:
- Add a `StatusBitsBase` class that represent generic status register and associated methods
- `StatusBitsBase` includes a mechanism to decode and encode bitfields
- In the instrument file, create a proper `Status` class derived from `StatusBitsBase`, by adding all the relevant fields and the functions to encode/decode.
-  Status information can be accessed as property, e.g. `inst.status.Format` (`inst `is an instance of an instrument)

See below some sketch implementation.

```
class StatusBitsBase(ctypes.BigEndianStructure):
    """
    A bitfield structure containing the assignments for the status decoding
    """
    _pack_ = 1
    _get_process_ = {}
    _set_process_ = {}

    @staticmethod
    def _convert_from_bcd(bcd):
        """Converts a bcd value to a decimal value

        :param value: The value to unpack from bcd
        :returns: The number in decimal form
        """
        place, decimal = 1, 0
        while bcd > 0:
            nibble = bcd & 0xF
            decimal += nibble * place
            bcd >>= 4
            place *= 10
        return decimal

    def __str__(self):
        """
        Returns a pretty formatted string showing the status of the instrument

        """
        ret_str = ""
        for field in self._fields_:
            ret_str = ret_str + f"{field[0]}: {hex(getattr(self, field[0]))}\n"

        return ret_str

    def __getattribute__(self, name):
        val = super().__getattribute__(name)
        if name == "fields":
            return val
        if name in self.fields():
            process = super().__getattribute__('_get_process_')
            if name in process:
                val = process[name](val)

        return val

    def __setattr__(self, name, value):
        val = super().__getattribute__(name)
        if name in self.fields():
            process = super().__getattribute__('_set_process_')
            if name in process:
                value = process[name](value)

        super().__setattr__(name, value)

    def fields(self):
        return [desc[0] for desc in super().__getattribute__('_fields_')]

class HPLegacyInstrument(Instrument):
    """
    Class for legacy HP instruments from the era before SPCI, based on `pymeasure.Instrument`

    """

    status_desc = StatusBitsBase # To be overriden by subclasses

    def __init__(self, adapter, name, **kwargs):
        super().__init__(
            adapter, name,
            includeSCPI=False,
            **kwargs,
        )

        self.status_bytes_count = ctypes.sizeof(self.status_desc)
        log.info(f"Initializing {self.name}")

.
.
.
    @property
    def status(self):
        """
        Returns an object representing the current status of the unit.

        """
        
        self.write("B")
        reply = self.adapter.read_bytes(self.status_bytes_count)
        return self.status_desc.from_buffer(reply)

```

in the instrument file, you could do something like this:
```
class Status(StatusBitsBase):
    """
    A bitfield structure containing the assignments for the status decoding
    """
    _pack_ = 1
    _fields_ = [
        # Byte 0: Function, Range and Number of Digits
        ("Format", c_uint8, 1),  # Bit 7
        ("SRQ", c_uint8, 3),  # bit 4..6
        ("trigger", c_uint8, 2),  # bit 2..3
        ("range", c_uint8, 2),  # bit 0..1
        # Byte 1 & 2:
        ("Number", c_uint16, 16),
        # Byte 1:
        # ("NRDGS_MSD", c_uint8, 4),
        # ("NRDGS_2SD", c_uint8, 4),
        # Byte 2:
        # ("NRDGS_3SD", c_uint8, 4),
        # ("NRDGS_LSD", c_uint8, 4),
        ("not_used", c_uint8, 4),
        ("Delay", c_uint32, 28),
        # Byte 3:
        # ("Not_Used", c_uint8, 4),
        # ("Delay_MSD", c_uint8, 4),
        # Byte 4:
        # ("Delay_2SD", c_uint8, 4),
        # ("Delay_3SD", c_uint8, 4),
        # Byte 5:
        # ("Delay_4SD", c_uint8, 4),
        # ("Delay_5SD", c_uint8, 4),
        # Byte 6:
        # ("Delay_6SD", c_uint8, 4),
        # ("Delay_LSD", c_uint8, 4),
        ]

    _get_porocess_ = {
        "SRQ" : None, # TBD
        "Number" : StatusBitsBase._convert_from_bcd,
        "Delay" : StatusBitsBase._convert_from_bcd
        }

    _set_porocess_ = {
        # TBD
        }
.
.
.
class HP3437A(HPLegacyInstrument):
    """Represents the Hewlett Packard 3737A system voltmeter
    and provides a high-level interface for interacting
    with the instrument.
    """

    status_desc = Status

    def __init__(self, resourceName, **kwargs):
        super().__init__(
            resourceName,
            "Hewlett-Packard HP3437A",
            **kwargs,
        )
        self.packed_bits = Packed_bits # Is this needed ?
        self.packed_bytes = self.bytefield_factory(PACKED_BYTES) # Is this needed ?
        self.packed_data = self.union_factory(self.packed_bytes, self.packed_bits) # Is this needed ?
        log.info("Initialized HP3437A")


```

I think also the `packed_*` data can be avoided in some way, but I am not sure.
For fields like `SRQ`, take into account that bitfields can be nested.

Sorry for the long comment, please let me know your opinion.

Reply:
@msmttchr Thank you very much for the constructive feedback.
I will try to implement the idea and hopefully we will see it working in the next days..

Reply:
@msmttchr  have started to implement the changes as proposed:
_hplegacydevices.py_:
```class HPLegacyInstrument(Instrument):
    """
    Class for legacy HP instruments from the era before SPCI, based on `pymeasure.Instrument`

    """
    status_desc = StatusBitsBase  # To be overriden by subclasses

    def __init__(self, adapter, name, status_bytes, status_bitfield, **kwargs):
        super().__init__(
            adapter, name,
            includeSCPI=False,
            **kwargs,
        )

        self.status_bytes_count = ctypes.sizeof(self.status_desc)
        self.status_bits = self.status_desc
        self.status_bytes = self.bytefield_factory(status_bytes)
        self.status_union = self.union_factory(self.status_bytes, self.status_bits)

        log.info(f"Initializing {self.name}")

    def write(self, command):
        if command == "B":
            self.adapter.connection.write("B", termination="")
        else:
            # super().write(command)
            self.adapter.connection.write(command)

    def values(self, command, **kwargs):
        if command == "B":
            self.write(command)
            return self.adapter.connection.read_raw(**kwargs)
        else:
            return super().values(command, **kwargs)

    @property
    def status(self):
        """
        Returns an object representing the current status of the unit.

        """
        # current_status = self.decode_status(self.fetch_status())
        # return current_status
        self.write("B")
        reply = self.adapter.read_bytes(self.status_bytes_count)
        return self.status_desc.from_buffer(reply).
.
.
.

```
for easier testing I have started working on the HP3478A (which is fan-less, so easier debugging in the living room)
_exerpt from HP3478a.py:_
```
STATUS_BYTES = 5


class Status(StatusBitsBase):
    """
    Support-Class with the bit assignments for the 5 status byte of the HP3478A
    """

    _fields_ = [
        # Byte 1: Function, Range and Number of Digits
        ("digits", c_uint8, 2),  # bit 0..1
        ("range", c_uint8, 3),  # bit 2..4
        ("function", c_uint8, 3),  # bit 5..7

        # Byte 2: Status Bits
        ("int_trig", c_uint8, 1),
        ("auto_range", c_uint8, 1),
        ("auto_zero", c_uint8, 1),
        ("fifty_hz", c_uint8, 1),
        ("front_rear", c_uint8, 1),
        ("cal_enable", c_uint8, 1),
        ("ext_trig", c_uint8, 1),
        ("res1", c_uint8, 1),

        # Byte 3: Serial Poll Mask (SRQ)
        ("SRQ_data_rdy", c_uint8, 1),
        ("res2", c_uint8, 1),
        ("SRQ_syntax_error", c_uint8, 1),
        ("SRQ_internal_error", c_uint8, 1),
        ("SRQ_front_panel", c_uint8, 1),
        ("SRQ_cal_error", c_uint8, 1),
        ("res3", c_uint8, 1),
        ("SRQ_PON", c_uint8, 1),

        # Byte 4: Error Information
        ("ERR_cal", c_uint8, 1),
        ("ERR_RAM", c_uint8, 1),
        ("ERR_ROM", c_uint8, 1),
        ("ERR_slope", c_uint8, 1),
        ("ERR_AD", c_uint8, 1),
        ("ERR_AD_Link", c_uint8, 1),
        ("res4", c_uint8, 1),
        ("res5", c_uint8, 1),

        # Byte 5: DAC Value
        ("DAC_value", c_uint8, 8),
    ]

    def _decode_trigger(self):
        i_trig = self.int_trig
        e_trig = self.ext_trig
        if i_trig == 0:
            if e_trig == 0:
                trigger_mode = "hold"
            else:
                trigger_mode = "external"
        else:
            trigger_mode = "internal"
        return trigger_mode

    _get_process_ = {
         "SRQ": None,  # TBD
         "Trigger": _decode_trigger,
         "Delay": StatusBitsBase._convert_from_bcd
         }

    # _set_process_ = {
    #     # TBD
    #     }


class HP3478A(HPLegacyInstrument):
    """ Represents the Hewlett Packard 3478A 5 1/2 digit multimeter
    and provides a high-level interface for interacting
    with the instrument.
    """
    status_desc = Status

    def __init__(self, resourceName, **kwargs):
        kwargs.setdefault('read_termination', '\r\n')
        kwargs.setdefault('send_end', True)
        super().__init__(
            resourceName,
            "Hewlett-Packard HP3478A",
            status_bytes=STATUS_BYTES,
            status_bitfield=self.status_desc,
            **kwargs,
        )
.
.
.

```

However when I now try to call `meter.status.Trigger` I get the following error:
```
In [5]: meter.status.Trigger
Traceback (most recent call last):

  Input In [5] in <cell line: 1>
    meter.status.Trigger

  File ~/builds/pymeasure/pymeasure/instruments/instrument.py:209 in __getattribute__
    return super().__getattribute__(name)

  File ~/builds/pymeasure/pymeasure/instruments/hp/hplegacyinstrument.py:159 in status
    return self.status_desc.from_buffer(reply)

TypeError: underlying buffer is not writable
```
Is this TypeError comming from the fact that I commented out the `_set_command`above?

Reply:
No, I don't think so. I am sorry I was not able to test my sketch code, but looking at the documentation the error should be fixed by updating the `status` property as follow:
```
    @property
    def status(self):
        """
        Returns an object representing the current status of the unit.

        """
        # current_status = self.decode_status(self.fetch_status())
        # return current_status
        self.write("B")
        reply = bytearray(self.adapter.read_bytes(self.status_bytes_count))
        return self.status_desc.from_buffer(reply).

```
On the other hand, after this fix you should be able to access all the bitfields defined in `Status` class (with optional decoding function), but `Trigger` is not one of them.
The way you have defined the decode_trigger function leads me to think that it should be a property similar to this (I changed `Trigger` to lower case to respect code style conventions):
```
  @property
  def trigger(self):
        status = self.status
        i_trig = status.int_trig
        e_trig = status.ext_trig
        if i_trig == 0:
            if e_trig == 0:
                trigger_mode = "hold"
            else:
                trigger_mode = "external"
        else:
            trigger_mode = "internal"
        return trigger_mode
 
```

Reply:
Hello,

The addition of the `bytearray(..)` helped, now the status field are accessible.